Lib/asyncio/tasks.py

```diff
@@ -564,62 +564,106 @@ async def _cancel_and_wait(fut):
         fut.remove_done_callback(cb)
 
 
-# This is *not* a @coroutine!  It is just an iterator (yielding Futures).
-def as_completed(fs, *, timeout=None):
-    """Return an iterator whose values are coroutines.
-
-    When waiting for the yielded coroutines you'll get the results (or
-    exceptions!) of the original Futures (or coroutines), in the order
-    in which and as soon as they complete.
+class _AsCompletedIterator:
+    """Doubles as an async iterator of as-completed tasks and futures
+    from a supplied set of awaitables and a plain iterator of
+    coroutines that resolve to results from the supplied awaitables as
+    their underlying tasks or futures complete.
+    """
+    def __init__(self, aws, timeout):
+        from .queues import Queue  # Import here to avoid circular import problem.
+        self._done = Queue()
+        self._timeout_handle = None
 
-    This differs from PEP 3148; the proper way to use this is:
+        loop = events.get_event_loop()
+        todo = {ensure_future(aw, loop=loop) for aw in set(aws)}
```

Comment:
@serhiy-storchaka, why change this from a `frozenset` to `set`?

Reply:
In general, it is faster to create a mutable collection from iterable than an immutable collection, because the latter does not have append/add methods, so you need to create a mutable collection first, and then convert it to immutable one.

But the current implementation of frozenset in CPython uses the same structure as set and its constructor is efficient as well as the set constructor. So currently there is no difference. It is just a habit.

Reply:
I prefer `frozenset` for these anonymous ones, but will leave it up to you. Thanks!